File: Queue/linkedlist.js
class linkedlistNode {
    data;
    next;
    constructor(_data) {
        this.data = _data;
        this.next = null;
    } 
}

class linkedlistIterator {
    currentNode;
    constructor(node){
        this.currentNode = node;
    }
    data(){
        if (this.currentNode==null) return null;
        return this.currentNode.data;
    }
    next(){
    if (this.currentNode != null){
        this.currentNode = this.currentNode.next;
    } return this;
    }
    
    current(){
        return this.currentNode;
    }
    
}

class linkedList{
    head;
    tail;
    constructor(){
        this.head = null;
        this.tail = null;
        this.length = 0;
        
    }
    
    begin(){
        const itr = new linkedlistIterator(this.head);
        return itr;
    }
    
    printList(){
        var printData = "";
        for(var itr=this.begin(); itr.current() != null; itr.next()){
            
            printData += itr.data() + " -> ";
        }
        
        console.log(printData);
    }
    
    find(_data){
        for(var itr=this.begin(); itr.current() !=null; itr.next()){
            if (itr.data() == _data){
                return itr.current();
            }
        }
        return null;
        
    }
    findParent(node){
        for(var itr=this.begin(); itr.current() !=null; itr.next()){
            if (itr.current().next == node){
                return itr.current();
            }
        }
        return null;
        
    }
    
    insertLast(_data){
        var newNode = new linkedlistNode(_data);
  
        if (this.head == null){
            this.head = newNode;
            this.tail = newNode;
        }   else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
    }
    
    insertAfter(node_data, _data ){
        var node = this.find(node_data);
        var newNode = new linkedlistNode(_data);
        newNode.next = node.next;
        node.next = newNode;
        
        if (newNode.next == null){
            this.tail = newNode;
        }
        this.length++;
    }
    insertBefore(node_data, _data ){
        const node = this.find(node_data);
        var newNode = new linkedlistNode(_data);
        newNode.next=node
        var parent = this.findParent(node);
        ;
        if (parent == null){
            this.head=newNode;
        }
        else {
            parent.next=newNode;
        }
        this.length++;
    }
    
    deleteNode(node_data){
        var node = this.find(node_data);
        if (node == null) return;
        if (this.head == this.tail){
            this.head == null;
            this.tail == null;
        } else if (this.head ==node){
            this.head = node.next;
        } else {
            var parent = this.findParent(node);
            if(this.tail = node){
                parent.next = null;
                this.tail = parent;
            }
            else {
                patent.next = node.next;
            }
            
        }
        this.length--;
    }
    size(){
        return this.length;
    }
    
    insertFirst(_data){
        var newNode = new linkedlistNode(_data);
        
        if(this.head == null) {
            this.head = newNode;
            this.tail = newNode;
        }
        else {
            newNode.next = this.head;
            this.head = newNode;
        }
        
        this.length++;
    }
    
    deleteHead(){
        if(this.head == null) return;
        this.head = this.head.next;
        this.length--;
        
    }
    
}


module.exports = {
    linkedList: linkedList,
    linkedlistNode: linkedlistNode,
    linkedlistIterator: linkedlistIterator
  };

File: Queue/linked_list.cpp
// Online C++ compiler to run C++ program online
#include <iostream>
using namespace std;


    
class linkedListNode
{ public:
    int data;
    linkedListNode *next;
    
    linkedListNode ( int _data){
        
        this->data = _data;
        this->next = NULL;
    }
};


class linkedListIterator {
    
    private:
        linkedListNode *currentNode;
        
    public:
    linkedListIterator(){ currentNode = NULL;}
    linkedListIterator(linkedListNode *node){ currentNode = node;}
    
    int data(){
        return this->currentNode->data;
    }
    linkedListIterator next(){
        this->currentNode = this->currentNode->next;
        return *this;
    }
    
    linkedListNode *current(){
        return this->currentNode;
    }
    
    
    
};
class linkedList{
    public: 
        linkedListNode *head = NULL;
        linkedListNode *tail = NULL;
        
        linkedListIterator begin(){
            linkedListIterator itr(this->head);
            return itr;
        }
        
        void printList(){
            for (linkedListIterator itr=this->begin(); itr.current()!= NULL; itr.next()){
                cout << itr.data() << " - > ";
            }
            cout << "\n";
        }
         
        
    void inserLast(int _data){
        linkedListNode *newNode = new linkedListNode(_data);
        
        if (this->head ==NULL){
            
            this->head = newNode;
            this->tail = newNode;
        }
        
        else{
            this->tail->next = newNode;
            this->tail = newNode;
            
        };
        
    };
    
    linkedListNode *find(int _data){
    for (linkedListIterator itr=this->begin(); itr.current()!= NULL; itr.next()){
        if (itr.data() == _data){
            return itr.current();
        }
            }
                
        return NULL;
    }
    
    linkedListNode *findParent(linkedListNode *node){
    for (linkedListIterator itr=this->begin(); itr.current()!= NULL; itr.next()){
        if (itr.current()->next == node){
            return itr.current();
        }
            }
                
        return NULL;
    }
    
    void insertAfter(linkedListNode *node, int _data){
        
        linkedListNode *newNode = new  linkedListNode(_data);
        newNode->next = node->next;
        node->next = newNode;
        
        if (newNode->next ==NULL){
            this->tail = newNode;
        }
    }
        
    void insertBefore(linkedListNode *node, int _data){
        
        linkedListNode *newNode = new  linkedListNode(_data);
        newNode->next = node;
        linkedListNode *parent = this->findParent(node);
        
        if (parent ==NULL){
            this->head = newNode;
        }
        else {
            parent->next = newNode;
        }
        
        
    };
    
    void deleteNode(linkedListNode *node){
  
        if(this->head == this->tail){
            this->head=NULL;
            this->tail= NULL;
        }
        else if (this->head==node){
            this->head = node->next;
        }
        else{
            linkedListNode *parentNode=this->findParent(node);
        
            if(this->tail ==node){
                this->tail = parentNode;
            }
            else{
                parentNode->next ==node->next;
            }
        }
        
        delete node;
        
    }
};

  int main() {
    linkedList *list = new linkedList();
    
    list->inserLast(1);
    list->inserLast(2);
    list->inserLast(3);
    
    list->printList();
    
    list->insertAfter( list->find(2), 98);
    
     list->printList();
     
    list->insertBefore( list->find(1), 55);
     list->printList();
     
    list->deleteNode( list->find(1));
     list->printList();
     
     cout << "head is " << list->head->data<<"\n";
}